bothan-core/src/store.rs

```diff
@@ -1,90 +1,256 @@
+        query_ids: QueryIDs,
+    ) -> Result<(), Error> {
+        let key = Key::QueryIDs {
+            source_id: source_id.as_ref(),
+        };
+
+        let serialized = bincode::serialize(&query_ids)?;
+        self.inner
+            .write()
+            .await
+            .db
+            .put(key.to_prefixed_bytes(), serialized)?;
+        Ok(())
+    }
+
+    async fn insert_query_ids<S, I>(&self, source_id: &S, ids: Vec<I>) -> Result<Vec<bool>, Error>
+    where
+        S: AsRef<str>,
+        I: Into<String>,
+    {
+        let mut query_ids = self.get_query_ids(&source_id).await?.unwrap_or_default();
+        let inserted = ids
+            .into_iter()
+            .map(|id| query_ids.insert(id.into()))
+            .collect();
+
+        self.set_query_ids(&source_id, query_ids).await?;
+        Ok(inserted)
+    }
+
+    async fn remove_query_ids<S, I>(&self, source_id: &S, ids: &[I]) -> Result<Vec<bool>, Error>
+    where
+        S: AsRef<str>,
+        I: AsRef<str>,
+    {
+        let mut current_set = self.get_query_ids(&source_id).await?.unwrap_or_default();
+        let removed = ids
+            .iter()
+            .map(|id| current_set.remove(id.as_ref()))
+            .collect();
+
+        self.set_query_ids(&source_id, current_set).await?;
+        Ok(removed)
+    }
+
+    async fn get_asset_info<S, I>(&self, source_id: &S, id: &I) -> Result<Option<AssetInfo>, Error>
+    where
+        S: AsRef<str>,
+        I: AsRef<str>,
+    {
+        let key = Key::AssetStore {
+            source_id: source_id.as_ref(),
+            id: id.as_ref(),
+        };
+
+        let serialized = self.inner.read().await.db.get(key.to_prefixed_bytes())?;
+        let asset_info = serialized
+            .map(|b| bincode::deserialize(b.as_slice()))
+            .transpose()?;
+        Ok(asset_info)
+    }
+
+    async fn insert_asset_info<S, I>(
+        &self,
+        source_id: &S,
+        id: &I,
+        asset_info: AssetInfo,
+    ) -> Result<(), Error>
+    where
+        S: AsRef<str>,
+        I: AsRef<str>,
+    {
+        let key = Key::AssetStore {
+            source_id: source_id.as_ref(),
+            id: id.as_ref(),
+        };
+
+        let serialized = bincode::serialize(&asset_info)?;
+        self.inner
+            .write()
+            .await
+            .db
+            .put(key.to_prefixed_bytes(), serialized)?;
+        Ok(())
+    }
+
+    async fn insert_asset_infos<S, I>(
+        &self,
+        source_id: &S,
+        assets: Vec<(I, AssetInfo)>,
+    ) -> Result<(), Error>
+    where
+        S: AsRef<str>,
+        I: AsRef<str>,
+    {
+        let inner = self.inner.write().await;
+        for (id, asset_info) in assets {
+            let key = Key::AssetStore {
+                source_id: source_id.as_ref(),
+                id: id.as_ref(),
+            };
+            let serialized = bincode::serialize(&asset_info)?;
+            inner.db.put(key.to_prefixed_bytes(), serialized)?;
```

Comment:
Why not call `insert_asset_info`?

Reply:
if we call insert_asset_info() it'll have to lock and release the lock on the inner O(n) times which isn't ideal for as acquiring the lock on RwLock isn't that efficient